### DISCUSSION TITLE
Twilio api call hanging in both sync and async path defs

### DISCUSSION THREAD

**Discussion by mike-guaranteed** [fastapi/fastapi#10585] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
You will need twilio creds

from twilio.rest import Client
from fastapi import FastAPI
import os

account_sid = os.getenv('TWILIO_ACCOUNT_SID')
auth_token = os.getenv('TWILIO_AUTH_TOKEN')
twilio_number = os.getenv('TWILIO_NUMBER')
messaging_service_sid = os.getenv('TWILIO_MESSAGING_SERVICE_SID')

class TwilioAPI:
    def __init__(self, phone_number: str):
        self.client = Client(account_sid, auth_token)
        self.phone_number = phone_number

    def send_sms(self, body: str):
        # we enter the function, but this never returns/resolves
        message = self.client.messages.create(
            messaging_service_sid=messaging_service_sid,
            body=body,
            to=self.phone_number,
        )
        return message.sid

app = FastAPI()

@app.post("/")
def send_message_or_whatever():
    dummy_message = "dummy message"
    twilio_api_client = TwilioAPI()
    message = twilio_api_client.send_sms(dummy_message)
    return {"message": message}
```

### Description

I have a FastAPI app (obvs) that has an endpoint which uses Twilio's Client to send a message sync.  When I attempt to send the client message like above, regardless of whether I use a regular sync path definition or use `async` and try to move the call to a separate thread, the server becomes unresponsive.   

I made a bare bones python file that just constructs the client and creates the message and it works fine so I do not suspect it is twilio itself. 
If I make a file change during this period the reloader freezes as well (I'm assuming since the server has become non-responsive). 

```
Python 3.11
fastapi==0.104.1
twilio==8.2.0
uvicorn==0.23.2
starlette==0.27.0
```

I am running the app locally like so (I've also called uvicorn directly from the command line):

```python
if __name__ == '__main__':
    uvicorn.run('app:app', reload=True, port=5002)
```

I have a router in a separate file and call app.include_router(<the_router>) in a builder function for the app. Here's the twilio client (we have our own lil wrapper):

```python
from twilio.rest import Client
...get env variables

class TwilioAPI:
    def __init__(self, phone_number: str):
        self.client = Client(account_sid, auth_token)
        self.phone_number = phone_number

    def send_sms(self, body: str):
        # we enter the function, but this never returns/resolves
        message = self.client.messages.create(
            messaging_service_sid=messaging_service_sid,
            body=body,
            to=self.phone_number,
        )
        return message.sid
```
The route in question looks like this:

```python
@router.post("/endpoint")
def send_message_or_whatever(input: Input):
    ...get data from input, construct message
    ...we hit our database here and this works
    twilio_api_client = TwilioAPI()
    twilio_api_client.send_sms(message) <--- this is where it goes sideways
    return stuff
```
All the examples I have found on twilio's own blog do something like

```python
@router.post('/endpoint')
async def do_something():
    client = twilio.rest.Client() # synchronous client
    client.messages.create(...create message params)
```

Stuff I have tried:
- using async and sync path definitions. Even though we are "waiting" on twilio in a sync function it shouldn't really matter? We wait for the db at other points which is a network call with no issue. Right now I don't even care if its not the most optimal thing for performance.

- when using async I have tried to use await asyncio.get_event_loop().run_in_executor(...) to no avail, nothing happens

- I tried to use fastapi's background task. It still gets stuck at `client.messages.create` (I am guessing this is a wrapper around `asyncio.to_thread` or `run_in_executor`)

- wrapping the call to `client.messages.create` in a try/except block.  It never exits the block and no error is ever thrown. 

My next course of action was to start rolling back versions of FastAPI and python.  

### Operating System

macOS

### FastAPI Version

0.104.1

### Pydantic Version

2.4.2

### Python Version

3.11

**Comment by jw-00000** [comment]
I think this issue is more likely due to how you use the Twilio library inside your FastAPI code, rather than something related to FastAPI itself.

First, are you also absolutely sure the right parameters get passed to the `self.client.messages.create` call? For instance, in your example code, you never initialize the `phone_number` that the `TwilioAPI` constructor expects: it should be `TwilioAPI("0123")` instead of `TwilioAPI()`.

Second, can you enable Twilio's logging ([see here on how](https://github.com/twilio/twilio-python#enable-debug-logging)) to check if an HTTP request actually gets made?

**Comment by mike-guaranteed** [comment]
Sorry for the delay, I appreciate the response.  Turns out it had something to do with running it locally on arm Macs.  I dockerized the app and its been running fine ever since.  I'm not exactly sure what the problem was but here's my little improperly formatted truth chart of asgi servers and their event loops:

`uvicorn` + `uvloop`:  ❌ 
`uvicorn` + `asyncio`: ❌ 
`hypercorn` + `asyncio`:  ✅
`hypercorn` + `uvloop`: ❌ 

I then ran the dockerized app with these combinations and they all worked, so I'm just using `uvicorn` + `uvloop`.
